context.go

```diff
@@ -2,107 +2,154 @@ package parens
 
 import (
 	"errors"
+	"fmt"
+	"sync"
 
 	"github.com/spy16/parens/value"
 )
 
-var _ Context = (*basicContext)(nil)
-
-const globalFrame = "<global>"
-
 // ErrNotFound is returned by Evaluator when a symbol resolution fails.
 var ErrNotFound = errors.New("not found")
 
-// Context encapsulates execution state.  It is passed to Evaluator in order to evaluate
-// a form within a given execution context.  The default Context implementation IS NOT
-// safe for concurrent use without external synchronization.
-type Context interface {
-	Push(StackFrame)
-	Pop() StackFrame
+// ConcurrentMap to store vars in a stack frame.
+type ConcurrentMap interface {
+	Load(string) (value.Any, bool)
+	Store(string, value.Any)
 
-	SetGlobal(name string, val value.Any)
-	Resolve(name string) (value.Any, error)
-
-	NewChild() Context
+	// Native map[string]value.Any.  Useful for ranging over values, e.g. when
+	// printing a stack frame.
+	Native() map[string]value.Any
 }
 
-// NewContext returns a global context using the default Context implementation.
-// The returned Context IS NOT safe for concurrent use without external synchronization.
-func NewContext() Context {
-	return newContext([]StackFrame{{Name: globalFrame}})
-}
+// MapFactory is a constructor for an implementation of ConcurrentMap.
+type MapFactory func() ConcurrentMap
 
 // StackFrame .
 type StackFrame struct {
```

Comment:
🤔 I am wondering, since we have made the concurrent-map customisable, is it still needed to expose this publicly?

Reply:
I'm 99% sure that we do.  The reason is that users may want to implement their own `Expr`s, which manipulate the stack via `Push` and `Pop`.  In such cases, they will need to work on an exported type.

Reply:
I think that makes sense. I was also considering instead of making the frames exposed and ask the user to push the frame, what if we made the `Push()` itself take frame related things as args and construct the stackFrame internally. but let me think about this a bit and get back to you... I plan to experiment with bunch of ideas this weekend. so should be to come back with something solid.

Reply:
>what if we made the Push() itself take frame related things as args and construct the stackFrame internally.

I considered this too, but then the story for `Pop` becomes a bit more complicated.  I don't really like the idea of multiple return values for code-style reasons.

Relatedly, I'd like to avoid making `StackFrame` an interface for performance reasons.

>but let me think about this a bit and get back to you... I plan to experiment with bunch of ideas this weekend. so should be able to come back with something solid.

Sounds like a plan 🙂.  Feel free to flag additional issues in the meantime.

Reply:
After thinking about this a bit more, I think we don't need to export this 😅. Reason is, since we are not really allowing customisation of how the stack itself is implemented, I don't think it leaves enough room for any optimisations even if we expose stack-frame and push/pop publicly since users can control only "when" push/pop happens but not how.. Since push/pop are expected to happen at specific places (i.e., when a function invocation is happening), even "when" is not much of a choice. So I think for now it's best to keep it as an internal detail.

Reply:
👍 Makes sense.  We can always think about exporting this in the future if someone asks for it.